contrib/babelfishpg_tsql/src/dbcmds.c

```diff
@@ -1231,3 +1259,343 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
+		return;
+	}
+
+	if (OidIsValid(get_role_oid(db_datawriter, true)))
+	{
+		ereport(LOG,
+				(errcode(ERRCODE_DUPLICATE_OBJECT),
+				 errmsg("role \"%s\" already exists. Please drop the role and restart upgrade.", db_datawriter)));
+		return;
+	}
+
+	datdba = get_role_oid("sysadmin", false);
+	check_can_set_role(GetSessionUserId(), datdba);
+
+	initStringInfo(&query);
+	appendStringInfo(&query, "CREATE ROLE dummy INHERIT; ");
+	appendStringInfo(&query, "CREATE ROLE dummy INHERIT; ");
+
+	stmt_list = raw_parser(query.data, RAW_PARSE_DEFAULT);
+	Assert(list_length(stmt_list) == 2);
+
+	/* Set current user to session user for create permissions */
+	prev_current_user = GetUserNameFromId(GetUserId(), false);
+	bbf_set_current_user("sysadmin");
+
+	old_dbid = get_cur_db_id();
+	old_dbname = get_cur_db_name();
+	set_cur_db(dbid, dbname);
+
+	PG_TRY();
+	{
+		/* Replace dummy elements in parsetree with real values */
+    	stmts = parsetree_nth_stmt(stmt_list, i++);
+    	update_CreateRoleStmt(stmts, db_datareader, NULL, NULL);
+
+    	stmts = parsetree_nth_stmt(stmt_list, i++);
+    	update_CreateRoleStmt(stmts, db_datawriter, NULL, NULL);
+
+		/* Run all subcommands */
+		foreach(parsetree_item, stmt_list)
+		{
+			Node		*stmt = ((RawStmt *) lfirst(parsetree_item))->stmt;
+			PlannedStmt *wrapper;
+
+			/* need to make a wrapper PlannedStmt */
+			wrapper = makeNode(PlannedStmt);
+			wrapper->commandType = CMD_UTILITY;
+			wrapper->canSetTag = false;
+			wrapper->utilityStmt = stmt;
+			wrapper->stmt_location = 0;
+			wrapper->stmt_len = 0;
+
+			/* do this step */
+			ProcessUtility(wrapper,
+						"(CREATE DATABASE ROLES) ",
+						false,
+						PROCESS_UTILITY_SUBCOMMAND,
+						NULL,
+						NULL,
+						None_Receiver,
+						NULL);
+		}
+		/* make sure later steps can see the object created here */
+		CommandCounterIncrement();
+
+		/* Grant permissions on all the schemas in a database to db_datareader/db_datawriter */
+		grant_perms_to_dbreader_dbwriter(dbid, db_datareader, db_datawriter);
+
+        /* Add entries to the catalog if not exists. */
+		if (!entry_exists_in_bbf_auth_ext(db_datareader))
+			add_to_bbf_authid_user_ext(db_datareader, "db_datareader", dbname, NULL, NULL, true, true, false);
+		if (!entry_exists_in_bbf_auth_ext(db_datawriter))
+			add_to_bbf_authid_user_ext(db_datawriter, "db_datawriter", dbname, NULL, NULL, true, true, false);
+	}
+	PG_FINALLY();
+	{
+		bbf_set_current_user(prev_current_user);
+		set_cur_db(old_dbid, old_dbname);
+	}
+	PG_END_TRY();
+	bbf_set_current_user(prev_current_user);
+	set_cur_db(old_dbid, old_dbname);
+}
+
+/*
+* This function is only being used for the purpose of the upgrade script to create
+* database roles db_datareader and db_datawriter for each database if the database
+* does not have the roles yet.
+*/
+PG_FUNCTION_INFO_V1(create_database_roles_for_all_dbs);
+Datum
+create_database_roles_for_all_dbs(PG_FUNCTION_ARGS)
```

Comment:
Again, this code is common b/w https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2970 and your PR and I have already reviewed this in the other PR so let's make sure to take that definition.